### PR TITLE
Add dionysus recipe

### PR DESCRIPTION
**Pull request by mrzv**

### CODE REVIEW

**Comment by dougalsutherland** [comment]
Hi @mrzv, thanks for the contribution!

It looks like you named the file `meta.yml`, but the system is looking for a file named `meta.yaml` -- hence why the linter complained, and why even though it looks like the CIs all passed they actually didn't do anything. :)

I'll put in some other comments about the recipe in a second, too.

File: recipes/dionysus/meta.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
+{% set name = "dionysus" %}
+{% set version = "2.0.2" %}
+{% set sha256 = "78c24da06fd273f0103ad4cb86331f5f4518d7f0445e2cecd0e325ca1a2d011f" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  fn: {{ name }}-{{ version }}.tar.gz
```
Comment (dougalsutherland) [inline recipes/dionysus/meta.yml]: We don't actually need the `fn` key for conda-forge, you can drop this.

File: recipes/dionysus/meta.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+build:
+    number: 0
```
Comment (dougalsutherland) [inline recipes/dionysus/meta.yml]: Could you use 2-space indentation here, to be consistent with the rest of the files / the rest of conda-forge? And in the rest of the file.

File: recipes/dionysus/meta.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    script: python setup.py install --single-version-externally-managed --record record.txt
+
+requirements:
+    build:
+        - python
+        - setuptools
+        - cmake
+        - boost
```
Comment (dougalsutherland) [inline recipes/dionysus/meta.yml]: Since you're compiling C code here, you should use the `toolchain` package as a build dependency.

**Comment by mrzv** [comment]
@dougalsutherland Thanks a lot! I was very confused what the problem was. I'll rename the file and fix the problems you point out.

**Comment by mrzv** [comment]
@dougalsutherland Is there a way to specify the minimum GCC version to use? CircleCI failed because it's trying to use GCC 4.8.2, which is too old.

**Comment by dougalsutherland** [comment]
Unfortunately conda-forge doesn't currently have support for newer versions of gcc, as far as I know. You're also going to get failures on Windows with python 2.7 then, which has to use MSVC 9.

You could try building with `clangdev`, which packages recent versions of clang for all platforms. Here are a few examples of recipes that do that: [`xgboost`](https://github.com/conda-forge/xgboost-feedstock/blob/524b141ded9d7157910fa992148d8669bbcccfcd/recipe/meta.yaml), [`lightgbm`](https://github.com/conda-forge/lightgbm-feedstock/blob/master/recipe/meta.yaml), [`pocl`](https://github.com/conda-forge/pocl-feedstock/blob/f8326295a3363ac5da6a71a97bb994b0fd3c514e/recipe/meta.yaml).

File: recipes/dionysus/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  run:
+    - python
+
+test:
+  imports:
+    - dionysus
+
+about:
+  home: http://mrzv.org/software/dionysus2
+  license: BSD 3-Clause
```
Comment (dougalsutherland) [inline recipes/dionysus/meta.yaml]: One more nitpick I missed earlier: should be `BSD-3-Clause`, with the extra dash.

**Comment by mrzv** [comment]
@dougalsutherland Thanks. I'll check those out. Is the solution for Windows to add `skip: True  # [win]`?

**Comment by dougalsutherland** [comment]
`clangdev` has Windows packages, but I'm not sure if it works with Python extensions.

You should probably not include clangdev on Mac (add a `# [not osx]` selector, or maybe `# [linux]`) since it seems to work without it. Looks like on Linux you'll have to give it a hint to use clangdev by setting `CC` / `CXX` before running python. You'll also probably have to explicitly include `libcxx` to get things like `<algorithm>`.

I can try fiddling with this later....

**Comment by dougalsutherland** [comment]
Also, you don't have to bump build numbers now, just once it's turned into a feedstock and actually published. There's nothing to conflict while it's still in `staged-recipes`. :)

**Comment by dougalsutherland** [comment]
Current Linux failure is I think because you need `libcxx` at runtime too, since it ends up using symbols in `libc++.so`.

File: recipes/dionysus/build.sh [hunk 5]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+export CXX=clang++
```
Comment (dougalsutherland) [inline recipes/dionysus/build.sh]: This should be in an `if [[ $(uname) == Linux ]]` or similar block. It happens to not mess stuff up on Mac apparently, but still better to stick to the default settings from `toolchain` if they work.

**Comment by dougalsutherland** [comment]
Windows builds aren't going to work now because there's no `bld.bat` or `build/script` entry....

**Comment by mrzv** [comment]
@dougalsutherland Thanks for the suggestions. I like your way (that lets me avoid having `build.sh`). For now my main problem is exactly what you say: the path to `libc++.so` is not linked into the shared library. It's one of these rpath issues that I've always struggled with. So I need to figure out how to convince cmake to use the full path during linking.

**Comment by dougalsutherland** [comment]
Oh, I don't think it's an RPATH problem (yet): it's just that `libc++.so` isn't even present, because you have `libcxx` as a build dependency but not a runtime one.

**Comment by mrzv** [comment]
Oh. I didn't even think of that. Interesting. Let me try it.

**Comment by dougalsutherland** [comment]
1. Sorry, I gave you a bad `CXXFLAGS`, need a dash in there. :)

2. Your `boost` dependency should probably be `boost-cpp`; `boost` includes boost:python, which is also bringing in numpy and stuff in a weird way.

**Comment by mrzv** [comment]
Thanks for catching that. Teaches me to copy things without looking. And thanks for the `boost` advice. I definitely don't need boost.python here.

File: recipes/dionysus/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+build:
+  number: {{ build_number }}
+  script: python setup.py install --single-version-externally-managed --record record.txt  # [not linux]
+  script: >
+    CC=clang CXX=clang++ CXXFLAGS=-stdlib=libc++ LDFLAGS=-lc++abi
+    python setup.py install --single-version-externally-managed --record record.txt  # [linux]
```
Comment (isuruf) [inline recipes/dionysus/meta.yaml]: Having multiple scripts gives a warning when reading as yaml. You can do

```
script: 
  - python setup.py install --single-version-externally-managed --record record.txt  # [not linux]
  - CC=clang CXX=clang++ CXXFLAGS=-stdlib=libc++ LDFLAGS=-lc++abi python setup.py install --single-version-externally-managed --record record.txt  # [linux]
```

**Comment by dougalsutherland** [comment]
Linux built! 🙌 

Mac failed, because I was being dumb with the `build/script` bit: that's not how selectors work. This should work instead:
```
build:
  number: {{ build_number }}
  script: >
    CC=clang CXX=clang++ CXXFLAGS=-stdlib=libc++ LDFLAGS=-lc++abi  # [linux]
    python setup.py install --single-version-externally-managed --record record.txt
```
(This way the environment variables are only set on Linux.)

Also, who knows whether Windows will work, but we'll find out eventually. :)

File: recipes/dionysus/meta.yaml (hunk 6 above)
Comment (dougalsutherland) [reply recipes/dionysus/meta.yaml]: @isuruf My suggestion in https://github.com/conda-forge/staged-recipes/pull/3976#issuecomment-330100756 should be okay too, though a little unusual, right?

Comment (isuruf) [reply recipes/dionysus/meta.yaml]: It's unsual, but if it works, sure.

**Comment by mrzv** [comment]
@isuruf @dougalsutherland Thanks for pointing the fixes. Yes, Linux is finally under control: great catch on libcxx missing from runtime.

I haven't used Windows since the last century, so I'll be amazed if it actually works.

**Comment by dougalsutherland** [comment]
Awesome.

Windows plan of attack (because it's late in my timezone and AppVeyor is slow):
- Wait and hope to see if the current recipe builds. :)
- If it doesn't, my guess is that you cannot build python extensions with clang on Windows (unless your python's been compiled with clang as well, which it hasn't). In that case, we can try to see if it builds with MSVC: just change the `[not osx]` selectors for `clangdev`/`libcxx` to `[linux]` and see what happens.
- Almost definitely, the Python 2.7 build will fail. Maybe, the 3.x builds will work: they use a much newer compiler. If the 3.x ones work and 2.7 fails, you can just add `skip: true  # [win and py27]` under `build`; if they all fail, just `skip: true  # [win]`.

The one other possible thing to add to the recipe would be a more thorough test that things built properly than just importing `dionysus`. This isn't totally necessary, because clearly the main failure mode of linkage problems is checked by the import, but writing a little `run_test.py` that uses the library to do something basic and quick might help catch possible future issues.

**Comment by mrzv** [comment]
Wonderful. Thanks for the detailed comments and all your help. We'll figure it out, one way or another.

**Comment by mrzv** [comment]
The Windows build didn't even try to use clang. The errors are very bizarre; I don't begin to guess what's causing them. As far as I can tell, all builds are using Visual Studio 14, so not much wiggle room there. My inclination is to disable the Windows build, unless there are more ideas to try. I'll wait till tomorrow before I do.

**Comment by dougalsutherland** [comment]
Yeah, Windows probably needs some hints like the `CC` variable to make it actually use clang, but I'm not sure exactly what those are there. I'm pretty sure it wouldn't have worked anyway for ABI issues. And yes, things were weird because apparently `clangdev` on Windows is only built with VC14, so then it installed VC14 versions of all the packages...anyway. If you're happy to not worry about Windows, let's not worry about Windows.

Some minor cleanup points:

- Can you change the `[not osx]` selectors to `[linux]`? Won't actually change anything but should be clearer.
- Can you change the `CXXFLAGS` setting to `CXXFLAGS="${CXXFLAGS} -stdlib=libc++"`, and similar for `LDFLAGS`, to avoid clobbering the default settings from `toolchain`?

**Comment by dougalsutherland** [comment]
@conda-forge/core, I don't think conda-forge has any packages yet compiling with `clangdev` on Linux. Is this okay / any gotchas to be concerned about?

**Comment by isuruf** [comment]
Compiling with `clangdev` is fine, but using `libc++` is not. It's because mixing packages built with `libc++` and `libstdc++` is not going to work.

**Comment by mrzv** [comment]
@isuruf Is `libstdc++` or whatever its equivalent is on Windows reasonably recent? Also, how do I tell it to use clang on Windows?

**Comment by mrzv** [comment]
@isuruf Sorry, I got confused about context. You are talking about Linux. Does this mean that what we have now won't work? Because of `libc++`.

**Comment by mrzv** [comment]
@dougalsutherland Thanks for the suggestions. I've made them.

**Comment by mrzv** [comment]
@isuruf If `libc++` is not Ok, how do people use C++11 code with Conda on Linux.

**Comment by isuruf** [comment]
I think it's fine as long as you are using this as a python package and don't have other C++ extensions link to this.

**Comment by jakirkham** [comment]
The system provided compiler on Linux already has C++11 support.

**Comment by mrzv** [comment]
@jakirkham If you are talking about GCC 4.8.2, its C++11 support is lacking.

**Comment by jakirkham** [comment]
What is it lacking for you?

**Comment by jakirkham** [comment]
According to this [table]( https://gcc.gnu.org/projects/cxx-status.html#cxx11 ) `gcc` 4.8.1 has full C++11 support. So this would extend to 4.8.2 as well.

**Comment by mrzv** [comment]
For example: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=57520
The first version of GCC that I know of, with which Dionysus doesn't run into some variation of this, is 5.4.

**Comment by mrzv** [comment]
"Full support" doesn't mean that it's not buggy.

**Comment by scopatz** [comment]
@mrzv, is this ready to go in? Discussions about C++11 support aside...

**Comment by mrzv** [comment]
@scopatz As far as I'm concerned, it's ready.

**Comment by dougalsutherland** [comment]
Outside of the C++11 issues I'm happy too.

**Pull request merged**